storey/targets.py

```diff
@@ -888,48 +922,87 @@ def _init(self):
+    ]:
+        fields_data = [_TDEngineFieldData(*raw) for raw in self._connection.query(f"DESCRIBE {table_name};")]
+        tags_schema = []
+        reg_cols_schema = []
+        for field_data in fields_data:
+            name = field_data.field
+            type_ = field_data.type
+            try:
+                if field_data.note == "TAG":
+                    tags_schema.append((name, self._tdengine_type_to_tag_func[type_]))
+                else:
+                    reg_cols_schema.append((name, self._tdengine_type_to_column_func[type_]))
+            except KeyError:
+                raise TDEngineTypeError(
+                    f"Unsupported value type '{type_}' of field '{name}' in {self.__class__.__name__}"
+                )
+        return tags_schema, reg_cols_schema
 
-    async def _emit(self, batch, batch_key, batch_time, batch_events, last_event_time=None):
-        with StringIO() as b:
-            b.write("INSERT INTO ")
-            if self._table:
-                b.write(self._table)
-            else:  # table is dynamic
-                b.write(batch_key)
+    def _get_sql_template(self) -> str:
+        with StringIO() as sql:
+            sql.write("INSERT INTO ?")
             if self._supertable:
-                b.write(" USING ")
-                b.write(self._supertable)
-                b.write(" TAGS (")
-                for column_index in range(self._number_of_tags):
-                    value = batch[0].get(self._columns[column_index], "NULL")
-                    b.write(self._sanitize_value(value))
-                    if column_index < self._number_of_tags - 1:
-                        b.write(",")
-                b.write(")")
-            b.write(" VALUES ")
-            for record in batch:
-                b.write("(")
-                for column_index in range(self._number_of_tags, len(self._columns)):
-                    value = record.get(self._columns[column_index], "NULL")
-                    b.write(self._sanitize_value(value))
-                    if column_index < len(self._columns) - 1:
-                        b.write(",")
-                b.write(") ")
-            b.write(";")
-            insert_statement = b.getvalue()
-        self._connection.execute(insert_statement)
+                sql.write(f" USING {self._supertable} TAGS {self._get_params_template(self._number_of_tags)}")
+            sql.write(f" VALUES {self._get_params_template(self._number_of_values)};")
+            return sql.getvalue()
+
+    @staticmethod
+    def _get_tags_from_event(
+        tags_schema: list[tuple[str, Callable[[Any], "taosws.PyTagView"]]], event: dict
+    ) -> list["taosws.PyTagView"]:
+        return [tag_func(event.get(tag_name)) for tag_name, tag_func in tags_schema]
+
+    @staticmethod
+    def _raw_value_to_value(value):
+        if isinstance(value, datetime.datetime):
+            return int(value.timestamp() * 1000)  # use millisecond precision
```

Comment:
I'd clarify to avoid implying that we have any other choice.
```suggestion
            return int(value.timestamp() * 1000)  # convert to milliseconds
```

Reply:
But we do - we could use micro-sec or nano-sec:
https://docs.tdengine.com/reference/taos-sql/data-type/#timestamp

https://docs.tdengine.com/cloud/taos-sql/data-type/#data-types:
> Default precision is millisecond, microsecond and nanosecond are also supported.

Reply:
I meant we don't have a choice at this point in the code :)

Reply:
This part does not create the table so it does not decide on precision.
However, the table's creator has the choice, but we only support ms at the moment.